src/main/java/com/bunq/sdk/http/ApiClient.java

```diff
@@ -106,163 +92,207 @@
   }
 
-  private URI determineFullUri(String uri, Map<String, String> params) throws URISyntaxException {
-    URIBuilder builder = new URIBuilder(apiContext.getBaseUri().toString() + uri);
+  private HttpUrl determineFullUri(String uri, Map<String, String> params) {
+    HttpUrl.Builder urlBuilder = new HttpUrl.Builder()
+        .scheme(SCHEME_HTTPS)
+        .host(apiContext.getBaseUri())
+        .addPathSegment(apiContext.getApiVersoin())
+        .addPathSegments(uri);
+
     SortedMap<String, String> paramsSorted = new TreeMap<>(params);
 
     for (Map.Entry<String, String> param : paramsSorted.entrySet()) {
-      builder.addParameter(param.getKey(), param.getValue());
+      urlBuilder.addQueryParameter(param.getKey(), param.getValue());
     }
 
-    return builder.build();
+    return urlBuilder.build();
   }
 
-  private CloseableHttpResponse executeRequest(HttpUriRequest request,
-      Map<String, String> customHeaders, String uri) throws IOException {
+  private Response executeRequest(BunqRequestBuilder request,
+                                  Map<String, String> customHeaders, String uri) throws IOException {
     if (!URIS_NOT_REQUIRING_ACTIVE_SESSION.contains(uri)) {
       apiContext.ensureSessionActive();
     }
 
     setHeaders(request, customHeaders);
 
-    return httpClient.execute(request);
+    return httpClient.newCall(request.build()).execute();
   }
 
-  private void setHeaders(HttpUriRequest httpEntity, Map<String, String> customHeaders) {
-    setDefaultHeaders(httpEntity);
-    setCustomHeaders(httpEntity, customHeaders);
-    setSessionHeaders(httpEntity);
+  private void setHeaders(BunqRequestBuilder requestBuilder, Map<String, String> customHeaders) {
+    setDefaultHeaders(requestBuilder);
+    setCustomHeaders(requestBuilder, customHeaders);
+    setSessionHeaders(requestBuilder);
   }
 
-  private void setDefaultHeaders(HttpUriRequest httpEntity) {
-    httpEntity.setHeader(HEADER_CACHE_CONTROL, CACHE_CONTROL_NONE);
-    httpEntity.setHeader(HEADER_USER_AGENT, getVersion());
-    httpEntity.setHeader(HEADER_LANGUAGE, LANGUAGE_EN_US);
-    httpEntity.setHeader(HEADER_REGION, REGION_NL_NL);
-    httpEntity.setHeader(HEADER_REQUEST_ID, UUID.randomUUID().toString());
-    httpEntity.setHeader(HEADER_GEOLOCATION, GEOLOCATION_ZERO);
+  private void setDefaultHeaders(Request.Builder httpEntity) {
+    httpEntity.addHeader(HEADER_CACHE_CONTROL, CACHE_CONTROL_NONE);
+    httpEntity.addHeader(HEADER_USER_AGENT, getVersion());
+    httpEntity.addHeader(HEADER_LANGUAGE, LANGUAGE_EN_US);
+    httpEntity.addHeader(HEADER_REGION, REGION_NL_NL);
+    httpEntity.addHeader(HEADER_REQUEST_ID, UUID.randomUUID().toString());
+    httpEntity.addHeader(HEADER_GEOLOCATION, GEOLOCATION_ZERO);
   }
 
-  private void setCustomHeaders(HttpUriRequest httpEntity, Map<String, String> customHeaders) {
+  private void setCustomHeaders(Request.Builder requestBuilder, Map<String, String> customHeaders) {
```

Comment:
Also here the formatting can be nicer.

Reply:
Hmm might be IDE settings. Auto formatter not working as expected :sweat_smile: